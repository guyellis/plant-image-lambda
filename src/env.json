{
"PLANT_IMAGE_COMPLETE": "fake-value"
}
