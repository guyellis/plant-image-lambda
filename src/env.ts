/* eslint-disable */
export default {
  "LALOG_LEVEL": "info",
  "LOGGLY_TOKEN": "fake-token",
  "PLANT_IMAGE_COMPLETE": "fake-value",
  "PLANT_IMAGE_HOST": "fake-host",
  "PLANT_IMAGE_PORT": "fake-port"
};
/* eslint-enable */
