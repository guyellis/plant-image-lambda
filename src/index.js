const AWS = require('aws-sdk');
const util = require('util');
const Logger = require('lalog');
const gm = require('gm');

const env = require('./env.json');
const pipeline = require('./outer');

/**
 * Entry point from Lambda call
 * @param {LambdaEvent} event
 * @param {Context} ctx
 */
async function handler(event, ctx) {
  process.env.LOGGLY_TOKEN = env.LOGGLY_TOKEN;
  Logger.setLevel(process.env.LALOG_LEVEL);

  const logger = Logger.create({
    serviceName: 'plant-image-lambda',
    moduleName: 'n/a',
    addTrackId: true,
  });

  logger.trace({
    msg: 'Reading options from event',
    event: util.inspect(event, { depth: 5 }),
  });

  try {
    const deps = {
      s3: new AWS.S3(),
      logger,
      gm: gm.subClass({
        imageMagick: true,
      }),
    };

    const req = {
      event,
      deps,
    };

    await pipeline(req);
    ctx.done();
  } catch (err) {
    logger.error({
      msg: 'Error in pipeline',
      event: util.inspect(event, { depth: 5 }),
      err,
    });
    ctx.done(err);
  }
}

module.exports = {
  handler,
};
