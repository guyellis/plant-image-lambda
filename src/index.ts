// eslint-disable-next-line import/no-unresolved
import { S3Event } from 'aws-lambda';
import AWS from 'aws-sdk';
import sharp from 'sharp';
import { pipeline } from './outer';

import { BasicRequest } from './outer-1-extract-from-event';
import { RequestDeps } from './types';
import { logger } from './logger';
import { getError } from './utils';

/**
 * Entry point from Lambda call
 */
async function handler(event: S3Event): Promise<void> {
  try {
    logger.info({
      event,
      msg: 'Reading options from event',
    });

    const deps: RequestDeps = {
      logger,
      s3: new AWS.S3(),
      sharp,
    };

    const req: BasicRequest = {
      deps,
      event,
    };

    await pipeline(req);
    return Promise.resolve();
  } catch (error) {
    const err = getError(error);
    logger.error({
      err,
      msg: 'Error in pipeline',
    });
    return Promise.reject(error);
  }
}

export {
  handler,
};
