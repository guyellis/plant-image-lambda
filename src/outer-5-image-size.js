'use strict';

const imageSizes = require('./image-sizes');

// #5
// data:
//   bucketName
//   key
//   fileName
//   imageType
//   s3Object
//   buffer
function getImageSize(req, next) {
  const gm = req.deps.gm;
  const data = req.data;
  console.log('o5 getImageSize');
  gm(data.buffer).size((err, size) => {
    console.log('o5 got size:', size);
    data.imageSize = size;
    data.sizes = imageSizes.calcSizes(size.width);
    console.log('o5 size done:', data.sizes);
    next(err, req);
  });
}

module.exports = {
  getImageSize,
};
