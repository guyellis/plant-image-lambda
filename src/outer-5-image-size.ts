import { Dimensions } from 'gm';

import { calcSizes } from './image-sizes';
import { ConvertToJpgResponse, ConvertToJpgData } from './outer-3-convert-to-jpg';
import { ImageSize } from './types';

export interface ImageSizeData extends ConvertToJpgData {
  imageSize: Dimensions;
  sizes: ImageSize[];
}

export interface ImageSizeResponse extends Omit<ConvertToJpgResponse, 'data'> {
  data: ImageSizeData;
}

// #5
// data:
//   bucketName
//   key
//   fileName
//   imageType
//   s3Object
//   buffer
export const getImageSize = async (
  req: Readonly<ConvertToJpgResponse>): Promise<Readonly<ImageSizeResponse>> => {
  const { data, deps: { gm, logger } } = req;
  const method = '5. getImageSize()';

  logger.trace({
    method,
  });

  return new Promise((resolve, reject) => {
    gm(data.buffer).size((err: Error|null, size: Dimensions) => {
      if (err) {
        logger.error({
          method,
          err,
        });
        return reject(err);
      }

      const nextData: ImageSizeData = {
        ...data,
        imageSize: size,
        sizes: calcSizes(size.width),
      };

      const response: ImageSizeResponse = {
        ...req,
        data: nextData,
      };

      logger.trace({
        msg: '5. getImageSize got size',
        size,
        sizes: nextData.sizes,
      });
      return resolve(response);
    });
  });
};
