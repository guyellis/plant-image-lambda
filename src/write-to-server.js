
const fetch = require('node-fetch');
const env = require('./env.json');

const {
  PLANT_IMAGE_HOST,
  PLANT_IMAGE_PORT,
  PLANT_IMAGE_COMPLETE,
} = env;

async function httpPost(req) {
  const {
    deps: {
      logger,
    },
    data: {
      s3Object: {
        Metadata: metadata,
      },
      sizes,
    },
  } = req;
  const { presets: { trackId } = {} } = logger;

  const putData = JSON.stringify({
    metadata,
    sizes,
    trackId, // Allows receiver to use same trackId for logging
  });

  const port = parseInt(PLANT_IMAGE_PORT, 10);
  const protocol = port === 443 ? 'https' : 'http';

  const url = `${protocol}://${PLANT_IMAGE_HOST}/api/image-complete?token=${PLANT_IMAGE_COMPLETE}`;
  const options = {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(putData),
    },
  };

  try {
    const response = await fetch(url, options);
    const logData = {
      msg: 'Image sizing metadata update sent',
      url,
      options,
      env,
      putData,
      status: response.status,
    };
    if (response.status === 200) {
      logger.trace(logData);
    } else {
      logger.error({
        ...logData,
        msg: `Unexpected response status ${response.status} in PUT call`,
      });
    }
  } catch (err) {
    logger.error({
      msg: 'Error sending image sizing metadata',
      url,
      options,
      env,
      putData,
    });
  }

  return req;
}

module.exports = httpPost;
