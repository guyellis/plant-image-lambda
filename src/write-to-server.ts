import env from './env';
import { ImageSizeResponse } from './outer-5-image-size';
import { ImageCompleteBody } from './types/image-lambda-types';

export const writeToServer = async (
  req: Readonly<ImageSizeResponse>,
): Promise<Readonly<Response | null>> => {
  const {
    PLANT_IMAGE_COMPLETE,
    PLANT_IMAGE_HOST,
    PLANT_IMAGE_PORT,
  } = env;

  const {
    deps: {
      logger,
    },
    data: {
      s3Object: {
        Metadata: metadata,
      },
      sizes,
    },
  } = req;

  logger.trace({ env, msg: 'writeToServer(): Started', sizes });
  const { presets } = logger;
  const { trackId = '' } = presets ?? {};

  const putData: ImageCompleteBody = {
    metadata,
    sizes,
    trackId, // Allows receiver to use same trackId for logging
  } as ImageCompleteBody; // TODO: Remove this cast

  logger.info({ msg: 'writeToServer(): putData', putData });

  const port = parseInt(PLANT_IMAGE_PORT, 10);
  const protocol = port === 443 ? 'https' : 'http';

  // When we're testing we are going to be setting the port to an arbitrary
  // high value that will map back to where our test server is running.
  // In this case we'll want to append :<port> to the host.
  const portSuffix = port !== 443 && port !== 80 ? `:${port}` : '';

  const url = `${protocol}://${PLANT_IMAGE_HOST}${portSuffix}/api/image-complete?token=${PLANT_IMAGE_COMPLETE}`;
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };
  const options: RequestInit = {
    body: JSON.stringify(putData),
    headers,
    method: 'PUT',
  };
  let response: Response | null = null;

  logger.trace({ msg: 'writeToServer(): About to call fetch()', options, url });

  try {
    response = await fetch(url, options);
    const logData = {
      env,
      msg: 'Image sizing metadata update sent and received',
      options,
      putData,
      status: response.status,
      url,
    };
    if (response.status === 200) {
      logger.info(logData);
    } else {
      logger.error({
        ...logData,
        msg: `Unexpected response status ${response.status} in PUT call`,
      });
    }
  } catch {
    logger.error({
      env,
      msg: 'Error sending image sizing metadata',
      options,
      putData,
      url,
    });
  }

  return response;
};
