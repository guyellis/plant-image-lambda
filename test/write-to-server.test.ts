import _ from 'lodash';

import env from '../src/env';
import { mockLogger, mockLoggerReset, makeFakeFetchResponse } from './helper';
import { writeToServer } from '../src/write-to-server';
import { ImageSizeResponse } from '../src/outer-5-image-size';

let status = 200;
let mockThrow = false;

global.fetch = jest.fn().mockImplementation((): Promise<Response> => {
  if (mockThrow) {
    throw new Error('fake-error');
  }
  return Promise.resolve(makeFakeFetchResponse(status));
});

describe('write-to-server', () => {
  beforeEach(() => {
    mockLoggerReset();
  });

  test('should log an error if fetch returns non-200', async () => {
    const req = {
      data: {
        s3Object: {
          Metadata: 'fake-metadata',
        },
        sizes: 'fake-sizes',
      },
      deps: {
        logger: mockLogger,
      },
    } as unknown as ImageSizeResponse;

    status = 400;

    const result = await writeToServer(req);

    expect(result).toMatchInlineSnapshot(`
      {
        "headers": {},
        "ok": true,
        "redirected": false,
        "status": 400,
        "statusText": "OK",
        "type": "default",
        "url": "fake-url",
      }
    `);
    expect(mockLogger.error).toHaveBeenCalledTimes(1);
  });

  test('should log an error if fetch throws', async () => {
    const req = {
      data: {
        s3Object: {
          Metadata: 'fake-metadata',
        },
        sizes: 'fake-sizes',
      },
      deps: {
        logger: mockLogger,
      },
    } as unknown as ImageSizeResponse;

    mockThrow = true;

    const result = await writeToServer(req);

    expect(result).toMatchInlineSnapshot('null');
    expect(mockLogger.error).toHaveBeenCalledTimes(1);
  });

  test('should use https instead of http', async () => {
    env.PLANT_IMAGE_PORT = '443';
    const req = {
      data: {
        s3Object: {
          Metadata: 'fake-metadata',
        },
        sizes: 'fake-sizes',
      },
      deps: {
        logger: mockLogger,
      },
    } as unknown as ImageSizeResponse;

    status = 200;
    mockThrow = false;

    const result = await writeToServer(req);

    expect(result).toMatchInlineSnapshot(`
      {
        "headers": {},
        "ok": true,
        "redirected": false,
        "status": 200,
        "statusText": "OK",
        "type": "default",
        "url": "fake-url",
      }
    `);
    expect(mockLogger.error).not.toHaveBeenCalled();
  });

  test('should send trackId if on presets', async () => {
    const logger = _.clone(mockLogger);
    logger.presets = {
      trackId: 'fake-track-id',
    };
    env.PLANT_IMAGE_PORT = '443';
    const req = {
      data: {
        s3Object: {
          Metadata: 'fake-metadata',
        },
        sizes: 'fake-sizes',
      },
      deps: {
        logger,
      },
    } as unknown as ImageSizeResponse;

    status = 200;
    mockThrow = false;

    const result = await writeToServer(req);
    const logTraceMock = logger.trace as jest.Mock;
    expect(result).toMatchInlineSnapshot(`
      {
        "headers": {},
        "ok": true,
        "redirected": false,
        "status": 200,
        "statusText": "OK",
        "type": "default",
        "url": "fake-url",
      }
    `);
    expect(logger.error).not.toHaveBeenCalled();
    expect(logTraceMock.mock.calls[0]).toMatchSnapshot();
  });
});
