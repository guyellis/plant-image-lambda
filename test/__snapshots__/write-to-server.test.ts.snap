// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`write-to-server should send trackId if on presets 2`] = `
[
  {
    "env": {
      "LALOG_LEVEL": "info",
      "LOGGLY_TOKEN": "fake-token",
      "PLANT_IMAGE_COMPLETE": "fake-value",
      "PLANT_IMAGE_HOST": "fake-host",
      "PLANT_IMAGE_PORT": "443",
    },
    "msg": "writeToServer(): Started",
    "sizes": "fake-sizes",
  },
]
`;
